### PR TITLE
Feature/provider

### PR DESCRIPTION
**Pull request by gumbaba-gs**
Provider-specific account configuration should be determined from a provider-independent source.

## Motivation and Context
 #128

## How Has This Been Tested?

Tested by executing hamlet query with debug option and verified the provider id, successfully.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions

- [x] None
closes hamlet-io/executor/#128

## Checklist:

- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] None of the above.

### CODE REVIEW

File: execution/contextTree.sh [hunk 1]
```diff
@@ -1218,9 +1218,9 @@ function upgrade_cmdb_repo_to_v1_3_0() {
   local -A account_mappings
   readarray -t account_files < <(find "${GENERATION_DATA_DIR}" -type f -name "account.json" )
   for account_file in "${account_files[@]}"; do
-    aws_id="$( jq -r '.Account.AWSId' <"${account_file}" )"
+    provider_id="$( jq -r '.Account.ProviderId' <"${account_file}" )"
```
Comment (ml019) [inline execution/contextTree.sh:1221]: Needs the check that it it is not null as per the logic below in setContext.

Looks like we would need to completely update the accounts-cmdb. Please make the logic fall back to the existing setup to give us time to sort out the transition, in line with the code in setContext.

**Review by ml019 — CHANGES_REQUESTED** [review]

**Comment by gumbaba-gs** [comment]
Updated setContext to meet the legacy configuration of AWSId and AzureId

File: execution/contextTree.sh [hunk 2]
```diff
@@ -1218,7 +1218,9 @@ function upgrade_cmdb_repo_to_v1_3_0() {
   local -A account_mappings
   readarray -t account_files < <(find "${GENERATION_DATA_DIR}" -type f -name "account.json" )
   for account_file in "${account_files[@]}"; do
-    provider_id="$( jq -r '.Account.ProviderId' <"${account_file}" )"
+    provider_id="$( jq -r '.Account.ProviderId | select(.!=null)' < "${account_file}" )"
+    [[ -e "${provider_id}" ]] && provider_id="$( jq -r '.Account.AWSId | select(.!=null)' < "${account_file}" )"
```
Comment (ml019) [inline execution/contextTree.sh:1222]: -e only applies to files. I think you need -z

File: execution/contextTree.sh [hunk 3]
```diff
@@ -1218,7 +1218,9 @@ function upgrade_cmdb_repo_to_v1_3_0() {
...
+    [[ -e "${provider_id}" ]] && provider_id="$( jq -r '.Account.AzureId | select(.!=null)' < "${account_file}" )"
```
Comment (ml019) [inline execution/contextTree.sh:1223]: same as above

**Review by ml019 — CHANGES_REQUESTED** [review]

File: execution/contextTree.sh (hunk 2 above)
Comment (gumbaba-gs) [reply execution/contextTree.sh]: @ml019 updated the flag with the commit 

313e9e5

File: execution/contextTree.sh (hunk 3 above)
Comment (gumbaba-gs) [reply execution/contextTree.sh]: @ml019 updated the flag with the commit 

313e9e5

File: execution/setStackContext.sh [hunk 4]
```diff
@@ -188,14 +188,14 @@ if [[ ! -f "${CF_DIR}/${TEMPLATE}" ]]; then
...
 # Permit renaming of stack files without affecting existing stack status
-if [[ -f "${CF_DIR}/${STACK}" && -n "${AWSID}" ]]; then
-    STACK_NAME=$(jq -r ".Stacks[0].StackName" <  "${CF_DIR}/${STACK}")
-fi
-
-# Permit renaming of stack files without affecting existing stack status
-if [[ -f "${CF_DIR}/${STACK}" && -n "${AZID}" ]]; then
-    STACK_NAME=$(jq -r ".properties.outputs.resourceGroup.value" <  "${CF_DIR}/${STACK}")
-fi
+case ${ACCOUNT_PROVIDER} in
```
Comment (ml019) [inline execution/setStackContext.sh:191]: Looks like we've lost the check that the stack already exists - this processing should only apply if the stack exists.

**Review by ml019 — CHANGES_REQUESTED** [review]

File: execution/setStackContext.sh (hunk 4 above)
Comment (gumbaba-gs) [reply execution/setStackContext.sh]: @ml019 

fixed to check for the stack. with commit #73386a7

File: execution/contextTree.sh [hunk 5]
```diff
@@ -1324,9 +1327,9 @@ function upgrade_cmdb_repo_to_v1_3_1() {
   local -A account_mappings
   readarray -t account_files < <(find "${GENERATION_DATA_DIR}" -type f -name "account.json" )
   for account_file in "${account_files[@]}"; do
```
Comment (ml019) [inline execution/contextTree.sh:1329]: This needs the same default ing logic as at line 1223/4.

**Review by ml019 — CHANGES_REQUESTED** [review]

**Comment by ml019** [comment]
TO actually get the account.json files updated, we will need a new upgrade. The code that is being modified is for past upgrades that most cmdbs have already had applied.

**Comment by RossMurr4y** [comment]
> TO actually get the account.json files updated, we will need a new upgrade. The code that is being modified is for past upgrades that most cmdbs have already had applied.

I think that would make a good follow-up ticket?

File: execution/contextTree.sh (hunk 5 above)
Comment (gumbaba-gs) [reply execution/contextTree.sh:1329]: @ml019 update the required with commit 25f2908

**Comment by RossMurr4y** [comment]
@ml019 are you happy a new CMDB upgrade to be raised as a new ticket or would you like it addressed here? Would we just increase he semver by 1 or is there a source for the version numbers we must align with?

**Review by ml019 — APPROVED** [review]
No issues

**Pull request merged**